### PR TITLE
Implement menvcfg

### PR DESCRIPTION
**Pull request by Timmmm**
This implements the m/senvcfg(h) CSRs. This CSR is used to enable/disable extensions and behaviours for lower privilege modes. Currently the only implemented bit is FIOM which affects how fences work.

It also affects how atomic memory accesses work in non-cacheable regions, but the model does not currently support PMAs so that can't easily be implemented.

Note there is a partial implementation of `menvcfg` in #137, so this should simplify that PR a little.

### CODE REVIEW

File: handwritten_support/riscv_extras_sequential.lem
Comment (arichardson) [inline handwritten_support/riscv_extras_sequential.lem:156]: This is hardcoded to true in the platform implementations, does it make any sense to add it there vs hardcoding it in sail?

Comment (Timmmm) [reply handwritten_support/riscv_extras_sequential.lem:156]: It just makes it configurable without changing the Sail code. Though you're right, I will add an emulator flag.

Maybe with a warning because it's supposed to be enabled if you support virtual memory and IIRC you can't disable that currently.

Comment (jrtc27) [reply handwritten_support/riscv_extras_sequential.lem:156]: Depends what privileged spec version you implement. We need to support all versions that people are making implementations for, so can't just go with the requirements of the latest version.

**Comment by allenjbaum** [comment]
I think you need to define precisely what you mean by "disabling virtual
memory" means exactly.
By my standards (low as they are), setting SATP.MODE to Bare disables
virtual memory.
Clearing MISA.S also disables virtual memory...

File: handwritten_support/riscv_extras_sequential.lem
Comment (Timmmm) [reply handwritten_support/riscv_extras_sequential.lem:156]: Good point - that's something I've been wondering about actually. What version of the spec does the model target?

If it is "all of them" I would have expected some kind of `has_spec(Priv1_12)` checks but those don't seem to exist. For example the very first thing I checked from the privileged spec changes:

> Changed MRET and SRET to clear mstatus.MPRV when leaving M-mode.

is just done unconditionally in the model:

      if   cur_privilege != Machine
      then mstatus->MPRV() = 0b0;

Another example: the 1.12 spec also introduces `mstatush` which is unconditionally implemented in the model.

Perhaps the intention is to add something `has_spec(Priv1_12)` but it just hasn't been done yet? Or maybe the plan for supporting multiple spec versions hasn't been fully formed... Any insight appreciated!

Anyway, I'll just add an emulator flag for now. :-)

File: model/riscv_sys_regs.sail
Comment (Timmmm) [inline model/riscv_sys_regs.sail:154]: This comment is wrong - should be "It must be supported if supervisor mode is implemented *AND* non-bare addressing modes are supported." I'll update it.

**Comment by Timmmm** [comment]
> I think you need to define precisely what you mean by "disabling virtual memory" means exactly.

Not sure what you are referring to here, but anyway here's the relevant bit of the spec:

> If S-mode is not supported, or **if satp.MODE is read-only zero (always Bare)**, the implementation may make FIOM read-only zero.

Now that I read it again, I think I misread it. The Sail model doesn't let you make `satp.MODE` be read-only zero but you can disable S-mode so that would be a legitimate way to also have FIOM disabled. I'll add an emulator flag.

File: c_emulator/riscv_sim.c
Comment (arichardson) [inline c_emulator/riscv_sim.c:302]: It's now possible to use long-only options (see e.g. OPT_TRACE_OUTPUT), I wouldn't use one of the few remaining short options for such a rarely used one.

Comment (Timmmm) [reply c_emulator/riscv_sim.c:302]: Ah thanks, I missed that! Looks like the C compiler could do with having all of its flags changed to match the OCaml one.

File: handwritten_support/riscv_extras_sequential.lem
Comment (Timmmm) [reply handwritten_support/riscv_extras_sequential.lem:156]: Flag added.

File: model/riscv_sys_regs.sail
Comment (arichardson) [inline model/riscv_sys_regs.sail:860]: Does FIOM need to be hardwired to zero if it's not enabled?

**Review by arichardson — COMMENTED** [review]
This looks good to me, just one question.

**Comment by allenjbaum** [comment]
It appears to be allowed but not required :
    Priv 3.1.18:   If S-mode is not supported, or if satp.MODE is read-only
zero (always Bare), the implementation *may* make FIOM read-only zero.
Otherwise, it looks like it must be RW

File: model/riscv_sys_regs.sail
Comment (Timmmm) [reply model/riscv_sys_regs.sail:860]: Good question! I think you're probably right - I'll double check.

Comment (Timmmm) [reply model/riscv_sys_regs.sail:860]: Yeah looks like I totally forgot to actually hook up the config flag to this code, oops! I'll fix it.

Comment (Timmmm) [reply model/riscv_sys_regs.sail]: Done.

**Comment by Timmmm** [comment]
@arichardson any chance of a final review? Pretty please? :-)

**Comment by arichardson** [comment]
> @arichardson any chance of a final review? Pretty please? :-)

I'm happy with the change, but you'll need it to be approved by one of the maintainers .

**Review by arichardson — APPROVED** [review]

**Review by billmcspadden-riscv — CHANGES_REQUESTED** [review]
Reconsider the command line argument name.

File: c_emulator/riscv_platform.c
Comment (Timmmm) [reply c_emulator/riscv_platform.c]: Agreed, that is a much better name! I've updated it everywhere.

Comment (billmcspadden-riscv) [reply c_emulator/riscv_platform.c]: I don't see the changes.  What am I missing?

Comment (Timmmm) [reply c_emulator/riscv_platform.c]: I think it's just because this comment stays linked to the first commit and gets marked "Outdated". I added a second commit with the rename. If you go to the "Files changed" tab at the top you should see it.

**Review by billmcspadden-riscv — APPROVED** [review]

**Comment by jrtc27** [comment]
Commits were not squashed before merging

**Comment by Timmmm** [comment]
> Commits were not squashed before merging

Sorry I didn't know the policy for this project, but I agree. Rebase merging like this is probably the worst of both worlds.

Personally I would disable `merge commits` and `rebase merging` in the project settings so you have to use `squash merge` (i.e. squash and rebase).

**Comment by jrtc27** [comment]
> > Commits were not squashed before merging
> 
> Sorry I didn't know the policy for this project, but I agree. Rebase merging like this is probably the worst of both worlds.
> 
> Personally I would disable `merge commits` and `rebase merging` in the project settings so you have to use `squash merge` (i.e. squash and rebase).

We use rebase merging in other projects extensively. It works well if people curate their commits in the PR, and allows you to preserve a patch series as a patch series, rather than squashing it into one. Squashing a PR only works when PRs are logically a single change. Both have their uses, people just need to be disciplined when they come to merge to either squash when appropriate or check that the commit series is sensible.

**Comment by Timmmm** [comment]
Doesn't it break `git bisect` though since only the last commit is actually tested by CI? I personally don't see much value in preserving history for changes that were never on `master`. Anyway I will squash in future!

**Pull request merged**
